test-data/unit/check-overloading.test

```diff
@@ -2191,20 +2191,54 @@ def f(x: A, y: Optional[B] = None) -> C: ...
 def f(x: A, z: Optional[C] = None) -> B: ...
 def f(x, y=None, z=None): ...
 
-reveal_type(f(A(), B()))
-reveal_type(f(A(), C()))
+reveal_type(f(A(), B()))  # E: Revealed type is '__main__.C'
+reveal_type(f(A(), C()))  # E: Revealed type is '__main__.B'
 
 arg: Union[B, C]
-reveal_type(f(A(), arg))
-reveal_type(f(A()))
+reveal_type(f(A(), arg))  # E: Revealed type is 'Union[__main__.C, __main__.B]'
+reveal_type(f(A()))       # E: Revealed type is '__main__.D'
 
 [builtins fixtures/tuple.pyi]
-[out]
-main:16: error: Revealed type is '__main__.C'
-main:17: error: Revealed type is '__main__.B'
-main:20: error: Revealed type is '__main__.C'
-main:20: error: Argument 2 to "f" has incompatible type "Union[B, C]"; expected "Optional[B]"
-main:21: error: Revealed type is '__main__.D'
+
+[case testOverloadInferUnionWithDifferingLengths]
+from typing import overload, Union
+
+class Parent: ...
+class Child(Parent): ...
+
+class A: ...
+class B: ...
+
+@overload
+def f(x: A) -> Child: ...
+@overload
+def f(x: B, y: B = B()) -> Parent: ...
+def f(*args): ...
+
+x: Union[A, B]
+f(x)       # E: Argument 1 to "f" has incompatible type "Union[A, B]"; expected "A"
+f(x, B())  # E: Argument 1 to "f" has incompatible type "Union[A, B]"; expected "B"
```

Comment:
I think ideally we could support something like this. Could you please add a TODO here? (I see you have a TODO in code, but I would prefer to have them both).